### ISSUE TITLE
ImportDeclaration changes content in onEnd, but does not update AST

### ISSUE THREAD

**Issue by jvanharn** [fuse-box/fuse-box#690]
I am currently writing an FuseBox plugin that automatically rewrites the **templateUrl** to an `template: require(...)` style import, based on the **ng2-fused** package.

I discovered while writing this plugin that the imports of the AST's that I had rewritten were not updated from the ES2015 style imports to commonjs imports, but the ones I had not touched *were* updated.

I then discovered `ImportDeclaration` changes the code (file.contents) in it's **onEnd** method, but does not update the analysis ast tree. This then results in me writing the unaltered code to the bundle, which doesnt work, because it contains the ES2015 style imports :(

I have worked around the issue by forcing it to analyze the file 2 times, but that does not seem like an ideal solution.

Applicable files:
- https://github.com/fuse-box/fuse-box/blob/a66da50853412b9c4b8ebc1ee154c973d4e47ce3/src/analysis/plugins/ImportDeclaration.ts#L58
- https://github.com/fuse-box/fuse-box/blob/491ea15836cc7eb6d8a86a9a401d39952bcc6bfd/src/analysis/FileAnalysis.ts#L131

**Comment by nchanged** [comment]
@jvanharn I think somebody made a very good plugin already
https://github.com/alex-klock/ng2-fused

The part you are referring to is when a module comes from `node_modules` untranspiled.  Set you target in tsconfig
` "module": "commonjs", `

I hope that helps

**Comment by nchanged** [comment]
@jvanharn any updates?

**Comment by jvanharn** [comment]
The ng2-fused plugin does not support ES2015 compiled plugins (e.g. the meta data is compiled as properties of the class, and the ng2-fused plugin looks for the magic methods used in commonjs).
So I fixed this, but I kept seeing the original code, without the imports replaced by fuse-box/commonjs imports. Then I realised this is done in the ImportDeclaration analysis plugin.

In short, the bug I describe is an ordering problem where plugins ask for the ast tree using the method described in the wiki (fileanalisis) the onNode and onEnd event methods. And the ImportDeclaration overwrites the contents of the file, but doesnt update the AST tree. So the plugin works with the old AST tree, makes changes to this pre Typescript compilation thing which makes the es2015 imports, commonjs imports, and you end up with ES2015 files in your bundle, which dont execute of course.
